dnsproxy: Improve regex pattern

odinuge opened the pull request:
_TL;DR:_ This is now V3 of https://github.com/cilium/cilium/pull/20214.

The full list of changes can be found in the commits (tho. this is based off https://github.com/cilium/cilium/pull/21288);
- Extract regex anchors to the beginning and end of the full pattern
- Remove the use of regex groups since we don't use them
- Short circuit matchPatterns with just a wildcard and match all fqdns
- Skip running `.MatchString` on wildcard regexes.

The overall perf change for all commits in this PR vs. https://github.com/cilium/cilium/pull/21288 (all without the lru cache, and ignore the time/op, since they are running in docker for mac 😢 ):

```
$ benchstat <pre-this-pr> <this-pr>
name                             old time/op          new time/op          delta
_perEPAllow_setPortRulesForID-7           77.4s ± 8%           83.0s ±17%     ~     (p=0.310 n=5+5)

name                             old B(HeapInUse)/op  new B(HeapInUse)/op  delta
_perEPAllow_setPortRulesForID-7           97.3M ± 1%           74.2M ± 1%  -23.75%  (p=0.008 n=5+5)

name                             old alloc/op         new alloc/op         delta
_perEPAllow_setPortRulesForID-7          78.6GB ± 0%          73.0GB ± 0%   -7.18%  (p=0.008 n=5+5)

name                             old allocs/op        new allocs/op        delta
_perEPAllow_setPortRulesForID-7            167M ± 0%            108M ± 0%  -35.37%  (p=0.008 n=5+5)
```
---

After this we plan doing;
- Switch from string concatenation to a byte buffer in `GeneratePattern`
- Consider the regex for matchPatterns + `map[string]struct{}` for matchNames.
- Replacing the `[.]` with `\.` to reduce the regex size
- Replacing the `*` replacement from `[-a-zA-Z0-9_]*` to `[^.]*`
- ...

--

- [ ] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [ ] All code is covered by unit and/or runtime tests where feasible.
- [ ] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [ ] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [ ] Provide a title or release-note blurb suitable for the release notes.
- [ ] Thanks for contributing!

<!-- Description of change -->

Fixes: #issue-number

```release-note
dnsproxy: Improve regex used for matching dns queries by reducing its complexity and size to save memory and speed up matching
```

christarazi reviewed (changes requested):
Thanks for the PR! It would be great if we can get benchmarking code added to the tests. There should already be a framework that tests at a high-level inside `pkg/fqdn/dnsproxy/proxy_test.go`. However, it would be good to add more piece-meal benchmarks corresponding to the code changes in this PR.

`Benchmark_perEPAllow_setPortRulesForID_large` is skipped by default, but if you remove the `t.Skip()` and add the CNPs as a yaml to `testdata/cnps-large.yaml` and run go test, you should be able to get some results.

christarazi commented on pkg/fqdn/dnsproxy/proxy.go:
```diff
@@ -176,7 +177,7 @@ func (p *DNSProxy) checkRestored(endpointID uint64, destPort uint16, destIP stri
 
 	for i := range ipRules {
-		if _, exists := ipRules[i].IPs[destIP]; (exists || ipRules[i].IPs == nil) && ipRules[i].Re.MatchString(name) {
+		if _, exists := ipRules[i].IPs[destIP]; (exists || ipRules[i].IPs == nil) && ipRules[i].Re.MatchString(Reverse(name)) {
```
We probably would want to reverse the `name` once here ideally rather than on each `ipRules`. We can pass in a reversed `name` to `checkRestored` or do the reverse before the for-loop.

christarazi commented on pkg/fqdn/dnsproxy/proxy.go:
```diff
@@ -552,7 +553,7 @@ func (p *DNSProxy) CheckAllowed(endpointID uint64, destPort uint16, destID ident
 	for selector, re := range epAllow {
 		// The port was matched in getPortRulesForID, above.
-		if selector.Selects(destID) && re.MatchString(name) {
+		if selector.Selects(destID) && re.MatchString(Reverse(name)) {
```
Same as above re: reversing once rather multiple times needlessly

christarazi commented on pkg/fqdn/dnsproxy/proxy.go:
```diff
@@ -910,18 +914,40 @@ func GetSelectorRegexMap(l7 policy.L7DataMap) (CachedSelectorREEntry, error) {
-				reStrings = append(reStrings, "("+dnsPatternAsRE+")")
+				dnsRuleName := FromFqdn(dnsRule.MatchName)
+				dnsPatternAsRE, _ := CustomToRegexp(Reverse(dnsRuleName))
+				reStrings = append(reStrings, dnsPatternAsRE)
```
Do the `matchName`s need to be reversed? From what I understand, reversing the string makes sense if it's a pattern which contains multi-char matches such as `*` or `.+` type matches, i.e. `*.*.svc.cluster.local` because that allows the regex engine to "exit early". To elaborate, every DNS name comparison against the regex that starts with `*` and `*` will force the engine to go through all possible matches upfront, rather than when necessary.

For example, take the pattern `*.*.svc.cluster.local` and a DNS name `foo.bar.svc.cluster.local`. Without reversing, the regex engine will first process `foo` and `bar` and they'll match against the corresponding `*`'s. However, if the DNS name instead was `foo.bar.something.else.local`, then only after the `*` checks, does the regex engine realize this DNS name doesn't match.

Let's say we take the reverse of the pattern `local.cluster.svc.*.*` and `foo.bar.something.else.local`. The first `local` part will pass as a basic string match (less expensive than wildcard presumably). The second part, `else`, matches against `cluster` and the engine returns false, whereas if it was `*` it would've return true at this part of the domain.

If it's a name, then I don't see any improvement related to the above which can be possible, so we are probably just wasting cycles reversing the string.

(I think it would be helpful to put this as a code comment if accurate to explain why we do this.)

odinuge replied:
Hey!

> [...]
> If it's a name, then I don't see any improvement related to the above which can be possible, so we are probably just wasting cycles reversing the string.

Yee, very much agree. Ref. https://github.com/cilium/cilium/pull/20396 I think we can keep it as is for now, since that PR will likely be merged first! 😄 Also, if we wanted to reverse matchPatterns and not matchNames, they would either have to be in separate regexes, or we would have had to run `Match` on the domain twice; one each way.

christarazi commented on pkg/fqdn/dnsproxy/proxy.go:
```diff
@@ -910,18 +914,40 @@ func GetSelectorRegexMap(l7 policy.L7DataMap) (CachedSelectorREEntry, error) {
+				// a wildcard rule is present together with a set of other rules.
+				// TODO do even need any validation at all here, since we will only ever get valid domains?
+				// eg do only "."
+				if isAllAllowed {
```
```suggestion
				dnsPatternAsRE, isWildcard := CustomToRegexp(Reverse(dnsPattern))
				// If the pattern is a wildcard, quit the regex calculation and insert a simple allow
				// all rule. This will reduce memory usage and improve performance significantly when
				// a wildcard rule is present together with a set of other rules.
				// TODO do even need any validation at all here, since we will only ever get valid domains?
				// eg do only "."
				if isWildcard {
```

christarazi replied:
Can you elaborate on the TODO?

> TODO do even need any validation at all here, since we will only ever get valid domains?

odinuge replied:
Yeee. My thought is that in case we hit a wildcard, we don't really care about the regex to begin with. in that case, can we just return `^.*$` or something to always short circuit the regex?

And thanks for the suggestion, that looks better!

christarazi commented on pkg/fqdn/dnsproxy/proxy.go:
```diff
@@ -910,18 +914,40 @@ func GetSelectorRegexMap(l7 policy.L7DataMap) (CachedSelectorREEntry, error) {
+		// The impact of sorting is negligible for a low number of rules, and the more rules, the more the
+		// benefit of lower memory usage, faster compilation and faster matching start being noticeable
+		// TODO test this in real life to see for performance improvements
+		sort.Strings(reStrings)
```
```suggestion
		if !isWildcard {
		  sort.Strings(reStrings)
		}
```

christarazi replied:
Actually, we might want to separate out this change from the rest in a different commit. That way we can get benchmark results with this PR and then with this PR + sorting, so we can easily tell what sort of improvement / regression sorting might have.

odinuge replied:
Yee, agree!

wow, this makes much more of a diff than expected @christarazi! good call!

First one without sorting, number two with sorting; look at the heap diff
```
root@43144d6cc6bb:/code# go test -tags privileged_tests -v -run '^$' -bench Benchmark_perEPAllow_setPortRulesForID_large -benchmem -benchtime 1x -memprofile memprofile.out ./pkg/fqdn/dnsproxy
goos: linux
goarch: amd64
pkg: github.com/cilium/cilium/pkg/fqdn/dnsproxy
cpu: Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
Benchmark_perEPAllow_setPortRulesForID_large
Before Setup (N=1,EPs=20,cache=128)
Alloc = 3 MiB   HeapInuse = 5 MiB       Sys = 19 MiB    NumGC = 6
Before Test (N=1,EPs=20,cache=128)
Alloc = 37 MiB  HeapInuse = 49 MiB      Sys = 980 MiB   NumGC = 18
After Test (N=1,EPs=20,cache=128)
Alloc = 752 MiB HeapInuse = 892 MiB     Sys = 1405 MiB  NumGC = 47
Benchmark_perEPAllow_setPortRulesForID_large-7                 1        10003844804 ns/op       5845248776 B/op 37991006 allocs/op
PASS
ok      github.com/cilium/cilium/pkg/fqdn/dnsproxy      16.137s
root@43144d6cc6bb:/code# go test -tags privileged_tests -v -run '^$' -bench Benchmark_perEPAllow_setPortRulesForID_large -benchmem -benchtime 1x -memprofile memprofile.out ./pkg/fqdn/dnsproxy
goos: linux
goarch: amd64
pkg: github.com/cilium/cilium/pkg/fqdn/dnsproxy
cpu: Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
Benchmark_perEPAllow_setPortRulesForID_large
Before Setup (N=1,EPs=20,cache=128)
Alloc = 3 MiB   HeapInuse = 5 MiB       Sys = 20 MiB    NumGC = 6
Before Test (N=1,EPs=20,cache=128)
Alloc = 38 MiB  HeapInuse = 49 MiB      Sys = 934 MiB   NumGC = 18
After Test (N=1,EPs=20,cache=128)
Alloc = 446 MiB HeapInuse = 586 MiB     Sys = 935 MiB   NumGC = 49
Benchmark_perEPAllow_setPortRulesForID_large-7                 1        10052384684 ns/op       4664795352 B/op 41661104 allocs/op
PASS
ok      github.com/cilium/cilium/pkg/fqdn/dnsproxy      14.822s
```

christarazi commented on pkg/fqdn/dnsproxy/proxy.go:
```diff
@@ -910,18 +914,40 @@ func GetSelectorRegexMap(l7 policy.L7DataMap) (CachedSelectorREEntry, error) {
+		// Explicitly add end and beginning of string markers to the beginning and end to avoid having them
+		// defined on a per dns rule basis. Also avoid wrapping each dns rule inside a group since we don't
+		// need to know where a given match occur. All those will also improve the performance of both the
+		// regex compilation and the regex matching
```
Makes sense as something to be part of the commit msg.

odinuge replied:
Yee, sounds good! Will do the final touches of the commit msg when we converge towards the desired state. 👌

christarazi commented on pkg/fqdn/dnsproxy/proxy.go:
```diff
@@ -910,18 +914,40 @@ func GetSelectorRegexMap(l7 policy.L7DataMap) (CachedSelectorREEntry, error) {
+		// need to know where a given match occur. All those will also improve the performance of both the
+		// regex compilation and the regex matching
+		// TODO figure out if we can/want use a separate key for the regex cache to avoid the sorting
+		// and extra logic in case its present in the cache
```
What do you have in mind as keys to the cache?

odinuge replied:
Lets discuss all this in https://github.com/cilium/cilium/pull/20396 😄

christarazi commented on pkg/fqdn/dnsproxy/proxy.go:
```diff
@@ -930,3 +956,47 @@ func GetSelectorRegexMap(l7 policy.L7DataMap) (CachedSelectorREEntry, error) {
 }
+
+// Removes the last trailing dot in case its an fqdn
+// TODO move to the proper location
```
These can probably be moved to the `pkg/fqdn/matchpattern` pkg as there exists similar code to this there.

odinuge replied:
Sounds good, thanks!

christarazi commented on pkg/fqdn/dnsproxy/proxy.go:
```diff
@@ -930,3 +956,47 @@ func GetSelectorRegexMap(l7 policy.L7DataMap) (CachedSelectorREEntry, error) {
+		utf8.EncodeRune(buf[size-start:], r)
+	}
+	return string(buf)
+}
```
Would be good to write a benchmark test for this function so that we can evaluate other implementations and compare them.

christarazi commented on pkg/fqdn/dnsproxy/proxy_test.go:
```diff
@@ -800,17 +800,19 @@ func (s *DNSProxyTestSuite) TestFullPathDependence(c *C) {
 		"54": [{
-			"Re":  "(^example[.]com[.]$)",
+			"Re":  "^[.](?:moc[.]elpmaxe)$",
 			"IPs": null
```
One thing we can do if we are worried about the readability of the tests is to construct this string with `fmt.Sprintf`. I don't think this is too big of a deal, specifically as we have a comment explaining why it looks unexpected.

christarazi reviewed (changes requested):
Thanks for the PR, this is a great step towards a massive performance improvement in the Agent!

I understand this is a WIP :), and I have several comments about the code, but overall the idea is sound!

odinuge commented on pkg/fqdn/dns/dns.go:
```diff
@@ -37,3 +37,13 @@ func FQDN(s string) string {
+
+// FromFQDN returns the fqdn without the trailing dot.
+// If s isn't fully qualified, it behaves as the identity function.
+func FromFQDN(s string) string {
```
Hmmm. What should we name this?? 🤔

odinuge-work committed:
dnsproxy: remove regex matching groups

This now wraps the overall regex in anchors and keep the distinct
rules/fqdns as normal non-anchored regexes. It also removes the regex
groups since we don't use them. This makes the final regex much smaller,
(the heap in the benchmark is ~20% smaller), and makes it more
performant when used for matching.

$ benchstat <pre-this-commit> <this-commit>
name                             old time/op          new time/op          delta
_perEPAllow_setPortRulesForID-7           77.4s ± 8%           84.8s ±21%     ~     (p=0.310 n=5+5)

name                             old B(HeapInUse)/op  new B(HeapInUse)/op  delta
_perEPAllow_setPortRulesForID-7           97.3M ± 1%           78.3M ± 1%  -19.51%  (p=0.008 n=5+5)

name                             old alloc/op         new alloc/op         delta
_perEPAllow_setPortRulesForID-7          78.6GB ± 0%          77.2GB ± 0%   -1.86%  (p=0.008 n=5+5)

name                             old allocs/op        new allocs/op        delta
_perEPAllow_setPortRulesForID-7            167M ± 0%            113M ± 0%  -32.47%  (p=0.008 n=5+5)

Signed-off-by: Odin Ugedal <ougedal@palantir.com>

odinuge-work committed:
dnsproxy: exit early in case ruleset contain wildcard

If there is a wildcard rule in the ruleset, return a regex matching
everything early. This will reduce the size in memory, and the perf of
the dnsproxy with such rules in general.

$ benchstat <pre-this-commit> <this-commit>
name                             old time/op          new time/op          delta
_perEPAllow_setPortRulesForID-7           84.8s ±21%           83.0s ±17%    ~     (p=0.690 n=5+5)

name                             old B(HeapInUse)/op  new B(HeapInUse)/op  delta
_perEPAllow_setPortRulesForID-7           78.3M ± 1%           74.2M ± 1%  -5.27%  (p=0.008 n=5+5)

name                             old alloc/op         new alloc/op         delta
_perEPAllow_setPortRulesForID-7          77.2GB ± 0%          73.0GB ± 0%  -5.42%  (p=0.008 n=5+5)

name                             old allocs/op        new allocs/op        delta
_perEPAllow_setPortRulesForID-7            113M ± 0%            108M ± 0%  -4.28%  (p=0.008 n=5+5)

Signed-off-by: Odin Ugedal <ougedal@palantir.com>

odinuge-work committed:
dnsproxy: short circuit regexes matching all input

If we know the regex match all input, we don't really need to run the
matching. This will reduce the overall memory usage and allocations in
this code path.

Signed-off-by: Odin Ugedal <ougedal@palantir.com>

schlosna commented on pkg/fqdn/matchpattern/matchpattern_test.go:
```diff
@@ -25,14 +25,28 @@ var _ = Suite(&MatchPatternTestSuite{})
-		reStr := ToRegexp(source)
+		reStr := ToAnchoredRegexp(source)
+		_, err := regexp.Compile(reStr)
+		c.Assert(err, IsNil, Commentf("Regexp generated from pattern %sis not valid", source))
```
```suggestion
		c.Assert(err, IsNil, Commentf("Regexp generated from pattern %s is not valid", source))
```

odinuge replied:
ooo, good catch, thanks!

schlosna commented on pkg/fqdn/matchpattern/matchpattern.go:
```diff
@@ -76,3 +85,28 @@ func ToRegexp(pattern string) string {
+	return pattern
+}
+
+func escapeRegexpCharacters(pattern string) string {
```
Should `ToAnchoredRegexp` use this function as well?

https://github.com/cilium/cilium/blob/d0d9b6c9cec745192704b8518b98d67d541f4a2e/pkg/fqdn/matchpattern/matchpattern.go#L78-L83

odinuge replied:
Yeah, that was the plan! Thanks, updated the PR now!

schlosna commented on pkg/fqdn/matchpattern/matchpattern_test.go:
```diff
@@ -25,14 +25,28 @@ var _ = Suite(&MatchPatternTestSuite{})
 	for source, target := range map[string]string{
 		"cilium.io.":   "^cilium[.]io[.]$",
 		"*.cilium.io.": "^" + allowedDNSCharsREGroup + "*[.]cilium[.]io[.]$",
 		"*":            "(^(" + allowedDNSCharsREGroup + "+[.])+$)|(^[.]$)",
```
Should this use non-capturing groups?

```suggestion
		"*":            "^(?:(?:" + allowedDNSCharsREGroup + `+\.)+|\.)$`,
```

odinuge replied:
Yeah, I agree. I think its outside the scope of this PR tho., and imo we should just make that regex match everything, ala. `^.*$` or something, since by design it means _match everything. And if so, we can skip the regex `MatchString` call as well.

christarazi commented on pkg/fqdn/dnsproxy/proxy.go:
```diff
@@ -1123,23 +1122,26 @@ func shouldCompressResponse(request, response *dns.Msg) bool {
-			dnsPatternAsRE := matchpattern.ToRegexp(dnsPattern)
-			reStrings = append(reStrings, "("+dnsPatternAsRE+")")
+			dnsPatternAsRE := matchpattern.ToUnAnchoredRegexp(dnsPattern)
+			if dnsPatternAsRE == matchpattern.MatchAllUnAnchoredPattern {
+				return matchpattern.MatchAllAnchoredPattern
+			}
```
In an offline sync, you mentioned this was awkward. Could you explain why in a code comment?

odinuge replied:
👋 

Yeah, this line effectively makes the `MatchAllUnAnchoredPattern` var only used here. Was thinking about making the `ToUnAnchoredRegexp` func return `string, bool` and `true` when its a wildcard, but this is just as verbose and easy reason about. I don't feel too much for one over the other.

Any thoughts?

christarazi replied:
Ah, seems straightforward enough as-is.

odinuge-work committed:
fqdn/matchpattern: Fx typo in test

Also switch to %q to get quotes around it as well.

Signed-off-by: Odin Ugedal <ougedal@palantir.com>

odinuge-work committed:
fqdn/matchpattern: reuse regexp char escape

Reuse escapeRegexpCharacters func that implement the same escaping to
avoid duplication.

Signed-off-by: Odin Ugedal <ougedal@palantir.com>

christarazi reviewed (approved):
One thing I just realized is that once this lands, will it change the behavior of existing policies? Is there a potential it might break on existing matchpatterns in user environments?
